integration-test/Analysis/FixtureUtils.hs

```diff
@@ -0,0 +1,178 @@
+  Command
+    { cmdName = "nix-shell"
+    , cmdArgs = ["-I"] <> [nixExpression] <> ["--run"] <> [cmdName cmd <> " " <> Text.intercalate " " (cmdArgs cmd)]
+    , cmdAllowErr = cmdAllowErr cmd
+    }
+decorateCmdWith LocalEnvironment cmd = cmd
+
+-- --------------------------------
+-- Analysis fixture test runner
+
+performDiscoveryAndAnalyses :: (Has (Lift IO) sig m, AnalyzeProject a, MonadFail m) => AnalysisTestFixture a -> m [(DiscoveredProject a, DependencyResults)]
+performDiscoveryAndAnalyses AnalysisTestFixture{..} = do
+  sendIO $ runCmd environment buildCmd
+  absoluteArtifactDir <- absoluteDirOf (extractAt artifact)
+  discoveryResult <- sendIO $ testRunnerWithLogger (discover absoluteArtifactDir) environment
+  case discoveryResult of
+    Left fb -> fail (show fb)
+    Right dps ->
+      forM dps $ \dp -> do
+        analysisResult <- sendIO $ testRunnerWithLogger (ignoreDebug $ analyzeProject (projectBuildTargets dp) (projectData dp)) environment
+        case analysisResult of
+          Left fb -> fail (show fb)
+          Right dr -> pure (dp, dr)
+  where
+    runCmd :: FixtureEnvironment -> Maybe (Command, Path Rel Dir) -> IO ()
+    runCmd env cmd =
+      case cmd of
+        Nothing -> pure ()
+        Just (c, dir) -> do
+          absDir <- absoluteDirOf dir
+          res <- runExecIOWithinEnv env $ exec absDir c
+          case res of
+            Left err -> fail (show err)
+            Right _ -> pure ()
+
+-- --------------------------------
+-- IO helpers for test runners
+
+absoluteDirOf :: Has (Lift IO) sig m => Path Rel Dir -> m (Path Abs Dir)
+absoluteDirOf relDir = sendIO $ PIO.makeAbsolute (analysisIntegrationCaseFixtureDir </> relDir)
+
+getArtifact :: Has (Lift IO) sig m => FixtureArtifact -> m (Path Abs Dir)
+getArtifact target = sendIO $ do
+  -- Ensure parent directory for fixture exists
+  PIO.ensureDir analysisIntegrationCaseFixtureDir
+  let artifactUrl = tarGzFileUrl target
+  let archiveExtractionDir = analysisIntegrationCaseFixtureDir </> extractAt target
+
+  PIO.ensureDir archiveExtractionDir
+  resolvedUrl <- useHttpsURI <$> mkURI artifactUrl
+
+  case resolvedUrl of
+    Nothing -> fail ("could not be resolved, artifact's download url: " <> show artifactUrl)
+    Just (url, _) -> do
+      PIO.withTempFile (analysisIntegrationCaseFixtureDir) "artifact" (downloadAndExtractArtifact url archiveExtractionDir)
+  
+  where
+    downloadAndExtractArtifact :: Url a -> Path Rel Dir -> Path Abs File -> Handle -> IO (Path Abs Dir)
+    downloadAndExtractArtifact url extractAt tempFile tempFileHandle = do
+      _ <- runReq defaultHttpConfig $
+        reqBr GET url NoReqBody mempty $
+        \r -> runConduitRes $ responseBodySource r .| CB.sinkFileCautious (toFilePath tempFile)
+
+      sendIO $ hClose tempFileHandle
+
+      sendIO $ PIO.ensureDir extractAt
+      archiveExtractFolder <- sendIO $ PIO.makeAbsolute extractAt
+      sendIO $ extractTarGz archiveExtractFolder tempFile
```

Comment:
We should probably support `extractTar` and `extractZip` here as well since we support those functions in the `UnpackArchives` logic.

Reply:
Leaving it for now, as we don't have zip in current PR, but I suspect we can get away without doing so, as GitHub supports getting .tar.gz of source code at commit, and release level.